### PR TITLE
Allow httplug 2.0

### PR DESCRIPTION
**Pull request by acrobat**
A pr to try/test if we can support both v1 and v2 of httplug. Currently there are still errors regarding typehint incompatibilities, so still a WIP to get things up and running

/cc @Nyholm is it actually possible to support both versions?

### CODE REVIEW

File: lib/Github/Client.php [hunk 1]
```diff
@@ -123,7 +123,7 @@ class Client
      */
     public function __construct(Builder $httpClientBuilder = null, $apiVersion = null, $enterpriseUrl = null)
     {
-        $this->responseHistory = new History();
+        $this->responseHistory = interface_exists('Http\Client\Common\HttpMethodsClientInterface') ? new Httplug2History() : new History();
```
Comment (GrahamCampbell) [inline lib/Github/Client.php]: You can just do `HttpMethodsClientInterface::class`. Importing a class that doesn't exist won't crash the runtime. It is purely just binding the name. No class loading will go on.

Comment (acrobat) [reply lib/Github/Client.php]: yes, you are right!

**Comment by acrobat** [comment]
> Yes, it is very possible.
> 
> Difference between version 1 and 2 of php-http/httplug is just PHP7 return types.


@Nyholm 

Ok bumping the supported php version to 7.1 (currently the oldest supported version) would kinda solve this. But if the failures in travis are related about return typehints, but adding them here (and matching those of httplug) is a BC break in our lib as users could extend these classes/methods. Or am I seeing thing wrong here?

**Comment by Nyholm** [comment]
> Ok bumping the supported php version to 7.1

Not needed if we still support version 1.0 of httplug.

**Comment by acrobat** [comment]
@Nyholm but without bumping php we can't support httplug 2.0. So I guess the only way to do it, is to bump the major version and switch to httplug v2?

**Comment by acrobat** [comment]
Ok, this looks promising! @Nyholm I've found your PR for the mailgun api lib and so I learned that the `VersionBridgePlugin` class exists. The only failing tests are now caused by a final class in httplug 2.0

I will take a look next time to get these tests passing again

**Comment by acrobat** [comment]
Implementation is done and all tests are green now, but as I suspected the [BC check fails now](https://travis-ci.org/KnpLabs/php-github-api/jobs/569520341). So I guess really the only way to do it is bump the httplug requirement to v2 in the next major release?

**Comment by Nyholm** [comment]
Sorry for being slow coming back to this. 

The BC warnings you see is actually fine =)

You get a bunch new from our `ExceptionInterface`. That is because HTTPlug starting to implement throwable. The `ExceptionInterface` has always intended to be implemented on an exception => which will implement these new functions... So you will only get issues if you have a non-exception and implement our `ExceptionInterface`. 

The other "BC warnings" comes from our plugins. They use the "version bridge" to automatically adapt if you installed HTTPlug v1 or v2.

**Comment by Nyholm** [comment]
Thank you @acrobat for the work you put in to this.

**Pull request merged**